torch_xla/csrc/helpers.cpp

```diff
@@ -481,11 +481,62 @@ xla::Shape XlaHelpers::GetPromotedShape(const xla::Shape& shape1,
 
 xla::Shape XlaHelpers::GetPromotedBinaryOpShape(const xla::Shape& shape1,
                                                 const xla::Shape& shape2) {
-  return xla::ShapeUtil::MakeShape(
-      PromoteType(shape1.element_type(), shape2.element_type()),
-      torch::lazy::GetPromotedShape(
-          xla::util::ToVector<int64_t>(shape1.dimensions()),
-          xla::util::ToVector<int64_t>(shape2.dimensions())));
+  if (!shape1.is_dynamic() && !shape2.is_dynamic()) {
+    return xla::ShapeUtil::MakeShape(
+        PromoteType(shape1.element_type(), shape2.element_type()),
+        torch::lazy::GetPromotedShape(
+            xla::util::ToVector<int64_t>(shape1.dimensions()),
+            xla::util::ToVector<int64_t>(shape2.dimensions())));
+  }
+  return GetPromotedDynamicShape(shape1, shape2);
+}
+
+xla::Shape XlaHelpers::GetPromotedDynamicShape(const xla::Shape& shape1,
+                                               const xla::Shape& shape2) {
+  std::vector<int64_t> upper_bounds1 =
+      xla::util::ToVector<int64_t>(shape1.dimensions());
+  std::vector<int64_t> upper_bounds2 =
+      xla::util::ToVector<int64_t>(shape2.dimensions());
+  absl::Span<const bool> dyn_dims1 = shape1.dynamic_dimensions();
+  absl::Span<const bool> dyn_dims2 = shape2.dynamic_dimensions();
+  std::vector<int64_t> upper_bounds;
+  std::vector<bool> dyn_dims;
+
+  if (upper_bounds1.size() > upper_bounds2.size()) {
+    upper_bounds.insert(
```

Comment:
some comment here would be nice. also what happen if `upper_bounds1.size() == upper_bounds2.size()`?

Reply:
If `upper_bounds1.size() == upper_bounds2.size()`, then line 512-516 should be no op. The python tests in this PR test this case.

Also added a comment for the broadcast rule.